panther_battery/include/panther_battery/battery_publisher.hpp

```diff
@@ -0,0 +1,182 @@
+  std::string error_msg_;
+
+  std::unique_ptr<panther_utils::MovingAverage<double>> battery_voltage_ma_;
+  std::unique_ptr<panther_utils::MovingAverage<double>> battery_temp_ma_;
+  std::unique_ptr<panther_utils::MovingAverage<double>> battery_current_ma_;
+  std::unique_ptr<panther_utils::MovingAverage<double>> battery_charge_ma_;
+};
+
+inline BatteryPublisher::BatteryPublisher(
+  rclcpp::Publisher<BatteryStateMsg>::SharedPtr battery_pub, const float & high_bat_temp,
+  const float & battery_charging_current_tresh, const int & batery_voltage_window_len,
+  const int & batery_temp_window_len, const int & batery_current_window_len,
+  const int & batery_charge_window_len, const float & battery_designed_capacity)
+: battery_pub_(std::move(battery_pub)),
+  high_bat_temp_(high_bat_temp),
+  battery_charging_current_tresh_(battery_charging_current_tresh),
+  battery_designed_capacity_(battery_designed_capacity)
+{
+  battery_voltage_ma_ = std::make_unique<panther_utils::MovingAverage<double>>(
+    batery_voltage_window_len, std::numeric_limits<double>::quiet_NaN());
+  battery_temp_ma_ = std::make_unique<panther_utils::MovingAverage<double>>(
+    batery_temp_window_len, std::numeric_limits<double>::quiet_NaN());
+  battery_current_ma_ = std::make_unique<panther_utils::MovingAverage<double>>(
+    batery_current_window_len, std::numeric_limits<double>::quiet_NaN());
+  battery_charge_ma_ = std::make_unique<panther_utils::MovingAverage<double>>(
+    batery_charge_window_len, std::numeric_limits<double>::quiet_NaN());
+}
+
+inline BatteryPublisher::~BatteryPublisher() { battery_pub_.reset(); }
+
+inline void BatteryPublisher::PublishUnknown(rclcpp::Time header_stamp)
+{
+  battery_voltage_ma_->Reset();
+  battery_temp_ma_->Reset();
+  battery_current_ma_->Reset();
+  battery_charge_ma_->Reset();
+
+  auto battery_msg = BatteryStateMsg();
+  battery_msg.header.stamp = header_stamp;
+  battery_msg.voltage = std::numeric_limits<double>::quiet_NaN();
+  battery_msg.temperature = std::numeric_limits<double>::quiet_NaN();
+  battery_msg.current = std::numeric_limits<double>::quiet_NaN();
+  battery_msg.percentage = std::numeric_limits<double>::quiet_NaN();
+  battery_msg.capacity = std::numeric_limits<double>::quiet_NaN();
+  battery_msg.design_capacity = battery_designed_capacity_;
+  battery_msg.charge = std::numeric_limits<double>::quiet_NaN();
+  battery_msg.cell_voltage = std::vector<float>(10, std::numeric_limits<float>::quiet_NaN());
+  battery_msg.cell_temperature = std::vector<float>(10, std::numeric_limits<float>::quiet_NaN());
+  battery_msg.power_supply_status = BatteryStateMsg::POWER_SUPPLY_STATUS_UNKNOWN;
+  battery_msg.power_supply_health = BatteryStateMsg::POWER_SUPPLY_HEALTH_UNKNOWN;
+  battery_msg.power_supply_technology = BatteryStateMsg::POWER_SUPPLY_TECHNOLOGY_LIPO;
+  battery_msg.present = true;
+  battery_msg.location = location_;
+
+  battery_pub_->publish(battery_msg);
+}
+
+inline void BatteryPublisher::Publish(
```

Comment:
There is something I don't like about that many parameters in the function. I might look better if there were for example setters for this. But it is up to discussion

Reply:
Had the same feeling. I replaced it with struct with params